### DISCUSSION TITLE
Rhizome on top of RDF

### DISCUSSION THREAD

**Discussion by andrewzhurov** [RhizomeDB/rs-rhizome#107] [Ideas]
Is there value from having RDF as the data model for Rhizome?
Or, perhaps any inspirations Rhizome can take from the Semantic Web ecosystem?

In particular, there is a graph querying framework, [Comunica](https://comunica.dev/), that's written in JS and can run in browsers.

On top of which there is [LDflex](https://github.com/LDflex/LDflex), that is a dot.notation interface on top of SPARQL, that aims to make it easy for web developers to interact with graph data. So they can write
`db.createUser({name: "Alice"})` to issue a SPARQL transaction,
`await db.user` to get data back,
`db.user.email.set("alice@alice.com")` to set an email and so on.
Perhaps something like this would be of value to give to Rhizome users to ease the learning curve?

There is also a GraphQL interface ([GraphQL-LD](https://github.com/rubensworks/graphql-ld.js)), for those that want the middle ground between ease of use of dot.notation and expressive power of SPARQL.

Another potential value that Semantic Web ecosystem has to offer is [RDF1.2](https://w3c.github.io/rdf-star/cg-spec/editors_draft.html), which landed just a couple of months back, and with it there is support for an edge to point to another edge, at last! (already [implemented in Comunica](https://github.com/comunica/comunica/pull/1222))
So edges/tuples can be accreated with any meta we wish, such as "Causality"/[which AST expression a tuple came from](https://github.com/RhizomeDB/rs-rhizome/discussions/106), temporal info and so on. (and we have access to this meta info from within a query, yay!)

Perhaps yet another value is to get the vision of Tim Berners-Lee, and his project [Solid](https://solidproject.org/), onboard.
He's after to make Semantic Web personal, allowing users to have their personal RDF stores (Solid PODs).
It's vision aligns with that of Rhizome.
Perhaps these two efforts can be aligned?
Having Solid on top of a robust data layer, such that of WNFS, would benefit it tremendously, imho.
That is what Rhyzome nailed dead right.
And having TBL onboard with Rhizome may give additional weight, social traction, and attract more developers into the Rhizome ecosystem.

I bet the ideas of the Semantic Web are known to Rhizome team.
Were any of them found to be of value and perhaps are planned to be adopted?
Perhaps Rhizome has a novel take on these ideas and already implements them in an even more powerful ways?

**Comment by expede** [comment]
Hi @andrewzhurov! 👋 

Thanks for writing this all up! I also need to get to your YouTube comment on the IPVM video :) I'm thrilled that you're so excited about this stuff!

> I bet the ideas of the Semantic Web are known to Rhizome team.

Yes, I got really into semantic web around ~2010 or so, and brush up on the latest updates in that community from time to time. We have a few RDF folks that comes to our community calls, as well. That said, we made intentional choice to not go down the RDF path before starting on the project.

We also didn't go down the path on the other side: strongly typed schemas. It's weird, because I'm a Haskeller at heart and believe in the power of types, but we're skating this middle path. The basic approach for the core of Rhizome is that [Worse is Better](https://dreamsongs.com/WIB.html), which is unsatisfying to those that have experienced how nice things can be without this tradeoff.

One way of looking at Rhizome is "RDF without the RDF" bits. There is definitely a segment of folks that are very into RDF. However, we've found that the majorioty of the devs we're hoping to serve aren't interested in these features, and are in fact turned off by the precieved complexity.

> Were any of them found to be of value and perhaps are planned to be adopted?

We use quads and SPOC for example — though we think about it more as EAVC instead of SPOC, which perhaps shows the subtle difference in appraoch. We don't use any of the "Resource Description" parts of RDF, and are focused on things like incremental view maintenance for collaborative docs, row-level encryption, and , which the RDF community has historically not needed / explored. Rhizome more focused on providing a database and CRDT primitives — despite the surface similarity of being a tuplestore, it's is closer to something like Datomic or Automerge than RDF.

RDF also has a strong association with the semantic web, which definitely still has a group of folks trying to push it forward, but hasn't managed to gain traction. Rhizome's approach is in many ways counter to the semamntic web ethos: unless you _want_ to impose order, let there be inconsistency!

> Perhaps Rhizome has a novel take on these ideas and already implements them in an even more powerful ways?

Rhizome is aimed at a related but distinct set of problems from RDF. It's also built in layers, so you could put whatever interface you wanted on it, including SPARQL (mroe below), but that's not in scope for the core project.

We have no current plans to deeply integrate with the RDF or linked data world at the core of Rhizome, but you could absolutely use Rhizome to build an RDF implementation, or implement a SPARQL-to-PomoLogic (the Rhizome IR) backend. We consider that all to be at the application and tooling layers — i.e. above what Rhizome itself provides.

That would be a seperate projects — a `rhizome-sparql` or `rhizome-rdf` or similar. Such projects are welcome, but we're really focused on the layers below that, and for a wider audience. It's all open source, so we couldn't stop people for doing so even if we wanted to!

**Comment by andrewzhurov** [comment]
Hi, @expede! Thank you for a detailed reply.
Pardon for absence. I find the problems that Rhizome & IPVM is onto solving to be deeply fascinating, they've been haunting me for some years. :smile:  I discovered there is a whole series on mindset behind Rhizome - Fission Reactor youtube series, and couldn't pass by this treasure trove of ideas, exploring them for the past week, for the sake of adopting insights and being more grounded in what Rhizome is and meant to be, in order to carry a more informed discussion (i.e., attempt to save myself from proposing straight out bollocks.:)

It's great to see that there are quite a few folks with extensive background in graph databases. (my exposure to Datomic and Semantic Web is rather shallow, compared to your collective intelligence, I hope you can forgive rookie mistakes I may make at times:)

Thanks for the link to the Worse is Better, [from my understanding](https://hyp.is/go?url=https%3A%2F%2Fdreamsongs.com%2FWIB.html&group=__world__) it's a design principle that values interface simplicity and implementation simplicity over the rest - build a solid simple core, increment it with features later - ideal design is the one you can't compact further, no the one with a ton of features. Seems to be a great design principle, resonates with me, a Clojure fan. Will serve Rhizome well.

> However, we've found that the majorioty of the devs we're hoping to serve aren't interested in these features, and are in fact turned off by the precieved complexity.

I see how Semantic Web as interface may be frightening to devs. That's a known hinder spot.
Same happens with Datalog as interface, although to a lesser degree.
More friendlier interfaces, that offer less learning curve, such as GraphQL and dot.notation seems to be of value to have.
But that's no news, well-known to you.:)
Perhaps GraphQL-LD and LDflex may serve as inspiration for similar interfaces on top of Rhizome.

> focused on things like incremental view maintenance

I see, incremental view maintenance is a powerful concept. Crucial to make CRDTs at Datalog-level efficient / practical.
I've taken a look at DRed and stumbled upon [Modular materialisation of Datalog programs](https://www.sciencedirect.com/science/article/pii/S0004370222000662?ref=pdf_download&fr=RR-2&rr=7f99c5ddc9827907) paper, whose approach is to build a dependency DAG of modules (for simplicity, say a module is a Datalog rule + backbone (state of rule inputs, per predicate), that is stratified. It seems Rhyzome adopts these two approaches, as it uses DBPS to propagate deltas, akin to DRed, through a DAG of rule dependencies (some of which may be wrapped under a stratum to manage non-monotonicity), caching rule's output (akin to keeping backbones, but in an immutable fashion (yay!)).
I heard Quinn say it's possible to turn a non-monotonic program into a chain/DAG of monotonic programs, or so I understood, but I struggle to imagine how it can be possible. It seems non-monotonicity is inherent in such program's design, so we only can try to manage it, e.g., with stratification / DAG of rule deps. Would love to learn where I'm confusing things.

I learned from Quinn's presentation that binary decision diagrams may be an interesting way to track provenance (what rules contribute to fact being true). That's one interesting concept! To me it seems as having rules as a reified AST. I.e., putting Dialog circuits to the level of DB and pointing from facts to circuits they've been produced from. Also resembles the design of Modular materialization (if only instead of stratums it used a DAG of dependencies that's reified).
The [paper](https://openproceedings.org/ICDT/2014/paper_36.pdf) I've taken look at also proposes usage of semirings for even more efficient representation of provenance, but didn't get to grokking on it yet, curious to learn your opinion on it, in case you happened to consider it.

In this context it's interesting to look at an open problem, at least as of April 2022, of [how to capture causality](https://youtu.be/P7X34y7xV9k?t=2220) [to input facts of a tx / facts that's been queried](https://youtu.be/P7X34y7xV9k?t=2461).
I.e., when issuing a Rhizome transaction txSomething(), how to capture causality to implicit facts that led to this tx being issued.
I'll try to illustrate the way I see this problem on example of:
Say you're shopping online, you think on buying a pair of nice headphones and see that there is only 2 available in stock, and there is a -40% discount that expires in 4 minutes! Somebody just bought one (Alice!), there's only the last one left.
You rush and add the last pair to your cart.
How would we capture causality that led to your decision? (only 1 in stock, -40% discount expiring in 4 mins)

A solution, perhaps, is to capture tx `addToCart(offer)` as a reified circuit / expression, that depends on other circuits that produced "in stock", "discount".

![image](https://github.com/RhizomeDB/rs-rhizome/assets/13056889/98be5cca-0548-4a83-acdb-0d252b8ae7ee)

Here arrows represent causal dependencies.
Nodes are expressions/functions/rules that derive new facts.
`Order` is yet another expression.

This DAG of expression seems akin to reified binary decision diagrams and Dialog's DBPS DAG.
Perhaps we can use the same model of DAG of expressions to serve for both provenance and incremental view maintenance?
I.e., have reified compute. Approach, described from another angle in [Rhyzome on top of IPVM](https://github.com/RhizomeDB/rs-rhizome/discussions/106).

This idea is based on my rather shallow understanding of all the mentioned tech, would love to learn why it may not make sense. :smile: 

---

EAV vs SPO.
From my understanding, the subtle difference between the two is in attempt of SPO to interlink mutable objects/places, whereas EAV aims to provide immutable and persistent values (entities).
Also SPO uses globally unique names in order to identified objects, whereas EAV makes use of content-based addressing.
Overall, I'm all up for EAV.
Because mutability makes a precarious foundation.
Also names make it difficult to interlink by content, there may exist multiple objects, describing the Moon, named differently (having different URLs), one may call it "the Moon", another "Luna" (spanish), another "Moon". We bravely buckle up and start interlinking names themselves ("sameAs" in RDF) but heck, wouldn't it be nice to not have the problem in the first place?
Names are baked into content. They seem to be valuable, as user-level concept, should we keep them there? (thanks to Unison lang for the insight)
RDF folks cleverly try to solve it with content-based addressing for names, however that creates a problem of propagating/superseeding new values, but I guess there is no escaping that.
So it's nice that EAV is name-less, akin to content-addressed RDF, making for a simpler model, allowing naming to exist, if one wishes, at a higher level.
Oh, well, that's all for the rant. :smile: Happy to learn your take on EAV vs SPO, perhaps I'm missing some bits.
Overall, immutable and persistent EAV seems to be a great foundation.
I love how it shines in combination with causality.
E.g.,
``` clojure
(setPrice 10 (createOffer {:product "Sick headhones" :price 10}))
```
Where `createOffer` creates an immutable entity
``` edn
{:product "Sick headphones"
 :price 10}
```
 and `setPrice` creates yet another immutable entity
``` edn
{:value   {:price 20}
 :depends {:product "Sick headphones"
           :price 10}}
```
each fact of which (which happened to be just one) points to the previous facts via Causality, allowing to have provenance on top of immutable persistent expressions and data.
Or, if we are to have binary decision diagrams, each fact points to dependent expressions (`createOffer`).
This example uses Clojure syntax to represent expressions and EDN to represent facts for I not know how Rhizome txes and triples are to be represented in text.

(the approach above (given it makes sense :smile: ) is just one way to have a mutable interface on top of immutable data, other's Datomic's txTime, XTDB's validTime)

Curious to learn your take on how a mutable interface with provenance can be achieved!

---

In keeping with the theme of EAV*C*, I'm curious how Causality can be expressed in general way in Rhizome.
I keep in mind the way of RDF*, which makes it possible to refer to triples.
*C* seems to allow for just that, but bakes this reference in (now) quads.
Is it feasible to have EAV + ability to address triples as a more generic approach?
There is one apparent problem with that - we don't have a DAG of causal dependencies anymore! :(
A -B-> C -D-> E
Where A represent an EAV (without `C`). B represents "A depends on C" and so on.
And so, A does not bake in it's causal dependency on B (that don't bake it's dependency on E).

Perhaps a way to mitigate it is to further extend the domain model to allow references to a set of CIDs.
Then we'd have
A -B-> #{C, C -D-> E}

:thinking: 

---

As for an RDF interface on top of Rhizome, now that I understand it better, regards to your explanation, perhaps Rhizome can be seen as a triple store.
If so, it seems possible to implement [RDF/JS Store](http://rdf.js.org/stream-spec/#store-interface) interface on top.
Which would make it possible to supply it as data source&sink to RDF graph querying frameworks, such as Comunica and Oxigraph.
If one wishes, that is.:)
It may offer RDF folks access to persistent and immutable data foundation, so may be of value to them.
However, I see that folks that use Rhizome directly may be with better DX by sticking with the interfaces tailored for Rhizome.

---

Oh, well, it ended up rather lengthy. :sweat_smile: I hope it's of value and at least somewhat close to "joy" to read. :smile: As always, would love to hear your thoughts.
